clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -4466,8 +4476,133 @@ std::unique_ptr<mlir::Pass> createConvertCIRToLLVMPass() {
+    moduleOp->walk(
+        [&](cir::StoreOp storeOp) { lowerTBAAAttr(storeOp, layout); });
+  }
+
+private:
+  template <typename Op> void lowerTBAAAttr(Op op, mlir::DataLayout &layout) {
+    auto tbaa = lowerTBAAToArrayAttr(op.getContext(), op.getTbaa(), layout);
+    op.setTbaaAttr(tbaa);
+  }
+
+  mlir::LLVM::TBAATagAttr lowerTBAAAttr(mlir::MLIRContext *ctx,
+                                        mlir::Attribute attr,
+                                        mlir::DataLayout &layout) {
+    if (auto scalarType =
+            mlir::dyn_cast<cir::TBAAScalarTypeDescriptorAttr>(attr)) {
+      auto accessType = lowerScalarType(ctx, scalarType, layout);
+      return mlir::LLVM::TBAATagAttr::get(accessType, accessType, 0);
+    }
+    llvm_unreachable("unknown tbaa type");
+  }
+
+  mlir::ArrayAttr lowerTBAAToArrayAttr(mlir::MLIRContext *ctx,
+                                       std::optional<mlir::ArrayAttr> tbaa,
+                                       mlir::DataLayout &layout) {
+    if (!tbaa) {
+      return mlir::ArrayAttr();
+    }
+    auto tbaaTagArray = *tbaa;
+    if (tbaaTagArray.empty()) {
+      return mlir::ArrayAttr();
+    }
+    llvm::SmallVector<mlir::Attribute, 2> tags;
+    for (auto attr : tbaaTagArray) {
+      auto tag = lowerTBAAAttr(ctx, attr, layout);
+      tags.push_back(tag);
+    }
+    return mlir::ArrayAttr::get(ctx, tags);
+  }
+
+  mlir::LLVM::TBAATypeDescriptorAttr
+  createScalarTypeNode(mlir::MLIRContext *ctx, llvm::StringRef typeName,
+                       mlir::LLVM::TBAANodeAttr parent) const {
+    llvm::SmallVector<mlir::LLVM::TBAAMemberAttr, 2> members;
+    members.push_back(mlir::LLVM::TBAAMemberAttr::get(ctx, parent, 0));
+
+    return mlir::LLVM::TBAATypeDescriptorAttr::get(
+        ctx, typeName, llvm::ArrayRef<mlir::LLVM::TBAAMemberAttr>(members));
+  }
+
+  mlir::LLVM::TBAARootAttr getRoot(mlir::MLIRContext *ctx) const {
+    return mlir::LLVM::TBAARootAttr::get(
+        ctx, mlir::StringAttr::get(ctx, "Simple C/C++ TBAA"));
+  }
+
+  mlir::LLVM::TBAATypeDescriptorAttr getChar(mlir::MLIRContext *ctx) const {
+    return createScalarTypeNode(ctx, "omnipotent char", getRoot(ctx));
+  }
+
+  mlir::LLVM::TBAATypeDescriptorAttr
+  lowerScalarType(mlir::MLIRContext *ctx,
+                  cir::TBAAScalarTypeDescriptorAttr scalarType,
+                  mlir::DataLayout &layout) const {
+    // special handle for omnipotent char
+    if (auto intTy =
+            mlir::dyn_cast_or_null<cir::IntType>(scalarType.getType())) {
+      if (intTy.getWidth() == 1 || intTy.getWidth() == 8) {
+        return getChar(ctx);
+      }
+    }
+    auto name = getTypeName(scalarType.getType());
+    return createScalarTypeNode(ctx, name, getChar(ctx));
+  }
+
+  StringRef getTypeName(mlir::Type type) const {
+    return TypeSwitch<mlir::Type, StringRef>(type)
+        .Case<cir::IntType>([](cir::IntType ty) { return getIntTypeName(ty); })
```

Comment:
Have you explored augmenting `CIR_Type` with another `let TBAAName = ...` name thing so that those can be done in tablegen and retrieved directly by an extra declaration function in tablegen?

Reply:
> augmenting CIR_Type with another let TBAAName = ... name

failed due to can't get the value of `TBAAName` in extra declaration function.

for example, test with

```
class CIR_Type <string name, string typeMnemonic> {
  let TBAAName = typeMnemonic;
  let extraClassDeclaration = [{
     llvm::StringRef getTBAATypeName() const {
       // ??? how to get the value of TBAAName
     }
  }]
}
```

Reply:
Sorry, what I meant is to write a custom tablegen driver to emit those, for example:
```
class UnaryFPToIntBuiltinOp<string mnemonic, string llvmOpName>
    : CIR_Op<mnemonic, [Pure]> {
  ...
  let llvmOp = llvmOpName;
}
```
Then `clang/utils/TableGen/CIRLoweringEmitter.cpp` looks at  `llvmOp` to populate the name of the intrinsic and facilitates some of the boilerplate for lowering 1-1 CIR to LLVM. Similarly, I'm suggesting you add a `let tbaaName` thing to `CIR_Type` and change tablegen to emit it for us, so that we can call a simple function to get the name.

I don't want to delay you further on this though, can you work on that as a follow up PR once we land this?

Reply:
gotcha. thanks for clarifying.
I will submit another patch to handle the `let tbaaName`.

Reply:
sent https://github.com/llvm/clangir/pull/1242